File: core/models/agent_config.py
"""全局 Agent 配置数据模型（分 Profile 目录 + registry.json）。"""

from typing import Literal

from pydantic import BaseModel, Field, model_validator

StyleVideoGenMode = Literal["text2video", "img2video", "keyframes"]


class CustomPromptProfile(BaseModel):
    """用户自定义 PromptProfile。"""

    id: str
    label: str
    based_on: str | None = None


class CustomStyleMode(BaseModel):
    """用户自定义或内置视频风格模式。"""

    id: str
    label: str
    default_prompt_profile: str = "default"
    include_video_gen: bool = False
    builtin: bool = False
    video: list[StyleVideoGenMode] | None = None

    @model_validator(mode="before")
    @classmethod
    def _migrate_legacy_include_video_gen(cls, data: object) -> object:
        """将旧版 include_video_gen 布尔字段迁移为 video 子模式列表。"""
        if not isinstance(data, dict):
            return data
        video = data.get("video")
        if video is None and data.get("include_video_gen"):
            data = dict(data)
            data["video"] = ["text2video", "img2video", "keyframes"]
        return data

    @model_validator(mode="after")
    def _sync_include_video_gen_flag(self) -> "CustomStyleMode":
        """include_video_gen 与 video 列表保持同步（API 兼容）。"""
        object.__setattr__(self, "include_video_gen", bool(self.video))
        return self


class AgentPromptContentOverride(BaseModel):
    """单 Agent 在某 profile 下的提示词覆盖。"""

    role_prompt: str | None = None
    action_hint: str | None = None


class AgentToolOverride(BaseModel):
    """单 Agent 工具白名单或黑名单（exclude 优先）。"""

    include_only: list[str] | None = None
    exclude: list[str] | None = None

    @model_validator(mode="after")
    def _normalize_lists(self) -> "AgentToolOverride":
        if self.include_only is not None:
            self.include_only = [x.strip() for x in self.include_only if x.strip()]
        if self.exclude is not None:
            self.exclude = [x.strip() for x in self.exclude if x.strip()]
        return self


class CustomAgentDefinition(BaseModel):
    """用户自定义 Agent（克隆内置子 Agent 能力）。"""

    id: str
    label: str
    based_on: str


class ProfileWorkspaceData(BaseModel):
    """单个 Profile 工作区配置（存于 profiles/{id}/workspace.json）。"""

    agent_roster: list[str] = Field(default_factory=list)
    custom_agents: list[CustomAgentDefinition] = Field(default_factory=list)
    prompt_content: dict[str, AgentPromptContentOverride] = Field(default_factory=dict)
    tool_overrides: dict[str, AgentToolOverride] = Field(default_factory=dict)


class AgentRegistryData(BaseModel):
    """全局 registry.json：Profile/风格元数据与全局 Agent 映射。"""

    custom_profiles: list[CustomPromptProfile] = Field(default_factory=list)
    style_modes: list[CustomStyleMode] = Field(default_factory=list)
    prompt_profiles: dict[str, str] = Field(default_factory=dict)
    tool_overrides: dict[str, AgentToolOverride] = Field(default_factory=dict)


class AgentConfigData(BaseModel):
    """聚合视图（API 兼容）：由 registry + 各 Profile 工作区组装。"""

    prompt_profiles: dict[str, str] = Field(default_factory=dict)
    custom_profiles: list[CustomPromptProfile] = Field(default_factory=list)
    style_modes: list[CustomStyleMode] = Field(default_factory=list)
    prompt_content: dict[str, dict[str, AgentPromptContentOverride]] = Field(
        default_factory=dict
    )
    tool_overrides: dict[str, AgentToolOverride] = Field(default_factory=dict)
    custom_agents: list[CustomAgentDefinition] = Field(default_factory=list)
    profile_agents: dict[str, list[str]] = Field(default_factory=dict)
    tool_overrides_by_profile: dict[str, dict[str, AgentToolOverride]] = Field(
        default_factory=dict
    )

File: apps/web/src/types/agentConfig.ts
/** Agent 配置 API 类型（提示词工作台） */

export interface PromptProfileOption {
  id: string;
  label: string;
  builtin?: boolean;
  deletable?: boolean;
  editable?: boolean;
  restorable?: boolean;
}

export interface CustomPromptProfile {
  id: string;
  label: string;
  based_on?: string | null;
}

export type StyleVideoGenMode = "text2video" | "img2video" | "keyframes";

/** JSON Schema 对象（工具入参/出参）。 */
export type JsonSchemaObject = Record<string, unknown>;

export interface StyleModeOption {
  id: string;
  label: string;
  default_prompt_profile: string;
  include_video_gen: boolean;
  /** 允许的 AI 生视频子模式；空表示不可 AI 生视频。 */
  video?: StyleVideoGenMode[];
  builtin: boolean;
}

export interface AgentPromptContentOverride {
  role_prompt?: string | null;
  action_hint?: string | null;
}

export interface AgentToolOverride {
  include_only?: string[] | null;
  exclude?: string[] | null;
}

export interface AgentToolInfo {
  name: string;
  description: string;
  action: string | null;
  read_only: boolean;
  kind?: string;
}

export interface CustomAgentDefinition {
  id: string;
  label: string;
  based_on: string;
}

export interface AgentToolOption {
  name: string;
  action: string;
  description: string;
  kind: string;
  read_only: boolean;
  scopes?: string[];
  operations?: string[];
  asset_layer?: string;
  affected_data_read?: string[];
  affected_data_write?: string[];
  boundary_note?: string;
  may_write_edit_timeline?: boolean;
  multi_scope_read?: boolean;
  /** 所属 Agent（全局 catalog 项）。 */
  agent?: string;
  /** LLM 调用入参 JSON Schema。 */
  input_schema?: JsonSchemaObject;
  /** 工具返回 observation JSON Schema。 */
  output_schema?: JsonSchemaObject;
}

export interface AgentInfo {
  name: string;
  display_name: string;
  builtin?: boolean;
  based_on?: string | null;
  action_pipeline: string[];
  ad_hoc_actions: string[];
  read_actions: string[];
  prompt_profile: string | null;
  effective_role_prompt: string;
  action_hint: string;
  tools: AgentToolInfo[];
  tool_options?: AgentToolOption[];
  system_tools?: AgentToolOption[];
  all_tools?: string[];
  effective_tools?: string[];
}

export interface AgentConfigResponse {
  prompt_profiles: Record<string, string>;
  custom_profiles: CustomPromptProfile[];
  style_modes: StyleModeOption[];
  prompt_content: Record<string, Record<string, AgentPromptContentOverride>>;
  tool_overrides: Record<string, AgentToolOverride>;
  custom_agents: CustomAgentDefinition[];
  profile_agents: Record<string, string[]>;
  tool_overrides_by_profile: Record<string, Record<string, AgentToolOverride>>;
  /** profile → agent → 可用 skill id；缺省 agent 键表示可用全部 */
  skill_allowlists_by_profile?: Record<string, Record<string, string[]>>;
  available_profiles: PromptProfileOption[];
  config_path?: string;
  agents: AgentInfo[];
}

export interface AgentConfigPatch {
  prompt_profiles?: Record<string, string>;
  custom_profiles?: CustomPromptProfile[];
  style_modes?: StyleModeOption[];
  prompt_content?: Record<string, Record<string, AgentPromptContentOverride>>;
  tool_overrides?: Record<string, AgentToolOverride>;
  custom_agents?: CustomAgentDefinition[];
  profile_agents?: Record<string, string[]>;
  tool_overrides_by_profile?: Record<string, Record<string, AgentToolOverride>>;
  skill_allowlists_by_profile?: Record<string, Record<string, string[]>>;
}

/** Skill 列表项（配置页 / Workbench） */
export interface SkillMetaItem {
  id: string;
  title: string;
  description?: string;
  aliases?: string[];
  /** 作用亮点（添加抽屉突出展示） */
  highlights?: string[];
  source?: string;
  deletable?: boolean;
}

export interface AgentPromptResponse {
  agent: string;
  profile: string;
  role_prompt: string;
  action_hint: string;
  source: { role_prompt: string; action_hint: string };
}

export interface ToolsCatalogResponse {
  governance?: {
    edit_timeline_write_agent: string;
    rules: string[];
  };
  catalog?: AgentToolOption[];
  agents: Record<string, AgentToolOption[]>;
  registry_version?: number;
}

export interface StyleModesResponse {
  style_modes: StyleModeOption[];
}
